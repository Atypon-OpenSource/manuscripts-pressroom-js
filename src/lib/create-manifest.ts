/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import builder from 'xmlbuilder'

interface ProcessingInstructions {
  priorityLevel: 'high'
  makeLiveCondition?: 'no-errors'
}

const buildProcessingInstructions = (
  processingInstructions: ProcessingInstructions
) => ({
  priority: {
    '@level': processingInstructions.priorityLevel,
  },
  // 'make-live': {
  //   '@on-condition': processingInstructions.makeLiveCondition,
  // },
})

interface ManifestOptions {
  groupDoi: string
  processingInstructions?: ProcessingInstructions
  submissionType: 'full' | 'partial'
}

export const buildManifest = ({
  groupDoi,
  processingInstructions,
  submissionType,
}: ManifestOptions): string => {
  const manifest = {
    submission: {
      '@dtd-version': '4.2',
      '@submission-type': submissionType,
      '@group-doi': groupDoi,
      'processing-instructions': processingInstructions
        ? buildProcessingInstructions(processingInstructions)
        : undefined,
    },
  }

  const options = {
    encoding: 'UTF-8',
    pubID:
      '-//Atypon//DTD Literatum Content Submission Manifest DTD v4.2 20140519//EN',
    sysID: 'manifest.4.2.dtd',
  }

  return builder.create(manifest, options).end({
    pretty: true,
  })
}
