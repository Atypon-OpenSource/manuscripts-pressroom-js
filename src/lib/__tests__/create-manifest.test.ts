/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { buildManifest } from '../create-manifest'

describe('manifest builder', () => {
  test('builds basic manifest', () => {
    const manifest = buildManifest({
      groupDoi: '10.0000/test',
      submissionType: 'full',
    })

    expect(manifest).toMatchSnapshot()
  })

  test('builds manifest with processing instructions', () => {
    const manifest = buildManifest({
      groupDoi: '10.0000/test',
      submissionType: 'partial',
      processingInstructions: {
        priorityLevel: 'high',
        makeLiveCondition: 'no-errors',
      },
    })

    expect(manifest).toMatchSnapshot()
  })
})
