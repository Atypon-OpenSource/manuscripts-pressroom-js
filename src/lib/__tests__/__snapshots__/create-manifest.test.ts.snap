// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manifest builder builds basic manifest 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE submission PUBLIC \\"-//Atypon//DTD Literatum Content Submission Manifest DTD v4.2 20140519//EN\\" \\"manifest.4.2.dtd\\">
<submission dtd-version=\\"4.2\\" submission-type=\\"full\\" group-doi=\\"10.0000/test\\"/>"
`;

exports[`manifest builder builds manifest with processing instructions 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE submission PUBLIC \\"-//Atypon//DTD Literatum Content Submission Manifest DTD v4.2 20140519//EN\\" \\"manifest.4.2.dtd\\">
<submission dtd-version=\\"4.2\\" submission-type=\\"partial\\" group-doi=\\"10.0000/test\\">
  <processing-instructions>
    <priority level=\\"high\\"/>
    <make-live on-condition=\\"no-errors\\"/>
  </processing-instructions>
</submission>"
`;
